File: backend/src/main/java/com/colegio/bastidas/service/AlumnoService.java
package com.colegio.bastidas.service;

import com.colegio.bastidas.dto.alumno.AlumnoCreadoResponseDTO;
import com.colegio.bastidas.dto.alumno.AlumnoRequestDTO;
import com.colegio.bastidas.dto.alumno.AlumnoResponseDTO;
import java.util.List;

/**
 * Contrato de negocio para la gestión de Alumnos.
 */
public interface AlumnoService {

    /** Lista todos los alumnos activos del año académico dado. */
    List<AlumnoResponseDTO> listarActivosPorAnio(Integer anio);

    /** Lista todos los alumnos de un aula específica. */
    List<AlumnoResponseDTO> listarPorAula(Long aulaId);

    /** Busca alumnos por nombre o DNI (búsqueda parcial). */
    List<AlumnoResponseDTO> buscar(String termino);

    /** Busca un alumno por su ID. */
    AlumnoResponseDTO buscarPorId(Long id);

    /** Busca un alumno por su DNI. */
    AlumnoResponseDTO buscarPorDni(String dni);

    /**
     * Crea un nuevo alumno y aprovisiona automáticamente su cuenta de Usuario
     * (rol ALUMNO, contraseña inicial predecible que debe cambiar en su primer login).
     */
    AlumnoCreadoResponseDTO crear(AlumnoRequestDTO dto);

    /** Actualiza datos de un alumno existente. */
    AlumnoResponseDTO actualizar(Long id, AlumnoRequestDTO dto);

    /** Cambia el estado de matrícula (ACTIVO, RETIRADO, TRASLADADO, EGRESADO) — baja lógica. */
    AlumnoResponseDTO actualizarEstado(Long id, String estado);

    /** Retorna la cantidad de alumnos activos en un año académico. */
    long contarActivosPorAnio(Integer anio);

    /**
     * Lista alumnos con permiso de academia activos.
     * Usado por el módulo de asistencia docente para mostrar el distintivo visual.
     */
    List<AlumnoResponseDTO> listarConPermisoAcademia(Integer anio);
}

File: frontend/src/app/core/services/alumno.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { CursoAsignado } from './docente.service';

export interface AlumnoResponse {
  id: number;
  nombres: string;
  apellidoPaterno: string;
  apellidoMaterno: string;
  nombreCompleto: string;
  codigoEstudiante: string;
  dni: string;
  fechaNacimiento: string;
  tienePermisoAcademia: boolean;
  horaEntradaAcademia?: string;
  activo: boolean;
  aulaId?: number;
  aulaDescripcion?: string;
  anioAcademico?: number;
  estadoMatricula?: 'ACTIVO' | 'RETIRADO' | 'TRASLADADO' | 'EGRESADO';
  nombreApoderado?: string;
  celularApoderado?: string;
  emailApoderado?: string;
  tutorId?: number;
  tutorNombre?: string;
}

export interface AlumnoRequestDTO {
  dni: string;
  apellidoPaterno: string;
  apellidoMaterno: string;
  nombres: string;
  fechaNacimiento?: string; // "YYYY-MM-DD"
  sexo?: 'MASCULINO' | 'FEMENINO';
  anioAcademico: number;
  aulaId?: number;
  tienePermisoAcademia?: boolean;
  horaEntradaAcademia?: string;
  nombreApoderado?: string;
  celularApoderado?: string;
  emailApoderado?: string;
  tutorId?: number;
}

export interface AlumnoCreadoResponse {
  alumno: AlumnoResponse;
  usernameGenerado: string;
  passwordInicial: string;
}

@Injectable({
  providedIn: 'root'
})
export class AlumnoService {
  private readonly BASE = `${environment.apiUrl}/alumnos`;

  constructor(private http: HttpClient) {}

  /**
   * Obtiene la lista de alumnos matriculados en una sección específica
   */
  listarPorAula(aulaId: number): Observable<AlumnoResponse[]> {
    return this.http.get<AlumnoResponse[]>(`${this.BASE}/aula/${aulaId}`);
  }

  /**
   * Busca un alumno por su DNI.
   */
  buscarPorDni(dni: string): Observable<AlumnoResponse> {
    return this.http.get<AlumnoResponse>(`${this.BASE}/dni/${dni}`);
  }

  /**
   * Busca un alumno por su ID.
   */
  buscarPorId(id: number): Observable<AlumnoResponse> {
    return this.http.get<AlumnoResponse>(`${this.BASE}/${id}`);
  }

  /**
   * Obtiene los cursos asignados al aula del alumno
   */
  obtenerCursos(id: number): Observable<CursoAsignado[]> {
    return this.http.get<CursoAsignado[]>(`${this.BASE}/${id}/cursos`);
  }

  /** Registra un nuevo alumno; el backend crea su cuenta de usuario automáticamente. */
  crear(dto: AlumnoRequestDTO): Observable<AlumnoCreadoResponse> {
    return this.http.post<AlumnoCreadoResponse>(this.BASE, dto);
  }

  /** Actualiza los datos de un alumno existente (no cambia DNI ni aula). */
  actualizar(id: number, dto: AlumnoRequestDTO): Observable<AlumnoResponse> {
    return this.http.put<AlumnoResponse>(`${this.BASE}/${id}`, dto);
  }

  /** Cambia el estado de matrícula (baja lógica): ACTIVO, RETIRADO, TRASLADADO, EGRESADO. */
  actualizarEstado(id: number, estado: string): Observable<AlumnoResponse> {
    return this.http.patch<AlumnoResponse>(`${this.BASE}/${id}/estado`, {}, { params: { estado } });
  }

  /** Lista los alumnos con permiso de academia pro-universitaria (Control de Ingresos Especiales). */
  listarConPermisoAcademia(anio: number): Observable<AlumnoResponse[]> {
    return this.http.get<AlumnoResponse[]>(`${this.BASE}/academia`, { params: { anio: String(anio) } });
  }
}
